### ISSUE TITLE
kernel connection issues

### ISSUE THREAD

**Issue by epsxy** [nteract/archived-desktop-app#2833]
#  Description

- It's nearly (see below) impossible to have a notebook working with an haskell kernel

# Reproduce

1. Installation of haskell
2. Installation of ihaskell
3. Try creating an haskell notebook with jupyter: `stack exec jupyter -- notebook` OK.
4. List every kernel installed OK: 
```
> jupyter kernelspec list --json
{
  "kernelspecs": {
    "haskell": {
        [...]
    },
    "python3": {
        [...]
    }
}
```
5. Launch nteract with haskell kernel: `nteract --kernel haskell` KO. Opens with bottom left `Python 3 | idle` and logs `Logging kernel python3`
6. Switch from Python runtime to Haskell KO. Go to Runtime > Select Haskell. Logs `Logging kernel haskell`. Bottom left is showing `Python 3 | not connected`. 
7. Restart Haskell runtime KO. Logs 
```
Kernel restarted
Kernel undefined has been restarted
ERROR: {}
```

Devtools: 
```
app.js:1 Error: launchKernelByNameEpic requires a kernel name
```

Everything looks like haskell kernel is properly detected by nteract (because it's proposed in the runtime selection menu) but I don't know why the kernel name is undefined in this context because it seems it's not the case when listing the kernels with the command line 4 above: `> jupyter kernelspec list --json`

# Why am I feeling confused?

1. Because my notebooks are working perfectly fine with Jupyter. 
2. I have try 2 different computers (OS.X Sierra 10.12.6 each), and try re-installing and ugrading  everything from haskell to python and nteract.
3. Because I have managed - no idea how - to make my notebooks work twice with nteract. I've seen one time the information `haskell | connected`, with all the snippet working fine. One time was: 
    - Not working before lunch
    - Locking computer
    - Working after lunch for a couple of minutes
4. When the haskell kernel was properly connected, closing and re-opening nteract would make my notebooks NOT working again (same error than explained above)

# Disclaimer

I am a Jupyter/Nteract beginner. Maybe I'm doing something wrong. There is not a lot of documentation and information about nteract online that's why it's possible I've missed some information.

Thanks for your help.

**Comment by epsxy** [comment]
I have finally managed to make my Haskell notebooks runs properly, but that's surprising. 

# Steps to reproduce
- Double click on your .ipynb
- nteract opens a new 'nteract instance' (i.e. a new 'windows') with at the bottom left: `Haskell | not connect`
- **Opens a second time** the **same** .ipynb
- nteract opens a new window with at the bottom left: `Haskell | not connect` 
- Most of the time, you will then find at the bottom left of the first window `Haskell | launched`
- If not, continue opening the notebook you want to run. After a few more openings, one of the window will probably have a Haskell kernel running. 

I've been trying this for a couple of minutes, and all the time, after 2 or 3 different 'nteract instance' of the same notebook, you'll have a haskell kernel running. 

# What should we do now?

I don't know if this issue is still relevant. There is probably something wrong. I've enabled the devtools and there is no warning or relevant information of what could have been wrong.

**Comment by rgbkrk** [comment]
Thanks for reporting this. I'm guessing that this is a result of our more recent handling of the kernel status and I should dive in with a slower to start kernel. We've been a bit blocked on a desktop release for a bit here, so this does need to get revisited.

**Comment by captainsafia** [comment]
Re-opening.

Thanks for this thorough report @epsxy! There have been a couple of releases of nteract since. Are you able to reproduce the issue in the latest release?

**Comment by epsxy** [comment]
Thanks for your answer @captainsafia! It's been a while I haven't used `nteract` for Haskell, and I had to restart the whole installation process for Haskell, IHaskell. Finally, *I can confirm I have been able to reproduce the issue*.

- Opening a Haskell notebook displays `no kernel | not connected`
- If I open a second time the same notebook, the kernel connects in the first one: `haskell | idle`, not in the second one I open

It looks as well pretty hard to change / select kernels under the menu `Runtime`. I don't think this problem is only related to Haskell though, I have had some troubles with python kernel connectivity as well.

Don't hesitate if I can do anything to help more!
